domain/personallearningplan/src/main/kotlin/uk/gov/justice/digital/hmpps/domain/personallearningplan/service/GoalService.kt

```diff
@@ -84,6 +100,73 @@ class GoalService(
       }
   }
 
+  /**
+   * Archives a [Goal], identified by its `prisonNumber` and `goalReference`, from the specified [ArchiveGoalDto].
+   */
+  fun archiveGoal(prisonNumber: String, archiveGoalDto: ArchiveGoalDto): Either<ArchiveGoalProblem, Goal> {
+    val goalReference = archiveGoalDto.reference
+    log.info { "Archiving Goal with reference [$goalReference] for prisoner [$prisonNumber] because [${archiveGoalDto.reason}] with description [${archiveGoalDto.reasonOther ?: ""}]" }
+
+    return goalPersistenceAdapter.getGoal(prisonNumber, goalReference).toOption()
+      .toEither { GoalToBeArchivedCouldNotBeFound(prisonNumber, goalReference) }
+      .flatMap { validateGoalIsValidForArchiving(it, prisonNumber, goalReference) }
+      .flatMap { validateArchiveDto(archiveGoalDto, prisonNumber, goalReference) }
+      .flatMap {
+        goalPersistenceAdapter.archiveGoal(prisonNumber, archiveGoalDto)?.right() ?: GoalToBeArchivedCouldNotBeFound(
+          prisonNumber,
+          goalReference,
+        ).left()
+      }
+  }
+
+  /**
+   * Unarchives a [Goal], identified by its `prisonNumber` and `goalReference`, from the specified [UnarchiveGoalDto].
+   */
+  fun unarchiveGoal(prisonNumber: String, unarchiveGoalDto: UnarchiveGoalDto): Either<UnarchiveGoalProblem, Goal> {
+    val goalReference = unarchiveGoalDto.reference
+    log.info { "Unarchiving Goal with reference [$goalReference] for prisoner [$prisonNumber]." }
+
+    return goalPersistenceAdapter.getGoal(prisonNumber, goalReference).toOption()
+      .toEither { GoalToBeUnarchivedCouldNotBeFound(prisonNumber, goalReference) }
+      .flatMap { validateGoalIsValidForUnarchiving(it, prisonNumber, goalReference) }
+      .flatMap {
+        goalPersistenceAdapter.unarchiveGoal(prisonNumber, unarchiveGoalDto)?.right() ?: GoalToBeUnarchivedCouldNotBeFound(
+          prisonNumber,
+          goalReference,
+        ).left()
+      }
+  }
+
+  private fun validateArchiveDto(
```

Comment:
I know that not having to declare the return types on kotlin functions is a big plus for the language, but I think in cases like these 3 private methods it would add some value, especially when reading the code outside of an IDE

I honestly can't tell you from reading the code what these 3 functions return. At a guess I would say an `Either<Goal, ArchiveGoalProblem>` ? 🤷‍♂️

Reply:
Close, it's the other way round because left = bad, right = good